File: internal/prompt/domain/prompt.go
package domain

import (
	"crypto/sha256"
	"encoding/hex"
	"time"
)

// PromptStatus represents the lifecycle stage of a prompt template.
type PromptStatus string

const (
	PromptDraft     PromptStatus = "draft"
	PromptPublished PromptStatus = "published"
	PromptArchived  PromptStatus = "archived"
)

// PromptTemplate is a versioned, content-addressed prompt.
// Once published it is immutable — new versions are created as new rows.
type PromptTemplate struct {
	Key         string       `json:"key"`       // system_prompt, memory_extraction, ...
	TenantID    *string      `json:"tenant_id"` // nil = global
	Version     int          `json:"version"`
	Content     string       `json:"content"`
	Status      PromptStatus `json:"status"`
	ContentHash string       `json:"content_hash"` // SHA-256 hex
	CreatedBy   string       `json:"created_by"`   // user:<id>|evolution:<cand_id>|system
	CreatedAt   time.Time    `json:"created_at"`
}

// PromptBinding wires a prompt key to specific versions for a scope
// (tenant or agent), with optional A/B traffic split.
type PromptBinding struct {
	Key             string `json:"key"`
	Scope           string `json:"scope"`             // tenant:<id>|agent:<id>
	StableVersionID string `json:"stable_version_id"` // always-served version
	CanaryVersionID string `json:"canary_version_id"` // experimental version
	TrafficPercent  int    `json:"traffic_percent"`   // 0-100, % routed to canary
}

// ComputeHash returns the SHA-256 hex digest of the content.
func ComputeHash(content string) string {
	h := sha256.Sum256([]byte(content))
	return hex.EncodeToString(h[:])
}

File: web/src/modules/prompt/model/prompt.ts
import { z } from 'zod';

// Prompt 模板列表行（GET /prompts）：每个 key 的最新版本摘要。
// 与 handler ListPrompts 渲染字段精确对齐。
export const promptSummarySchema = z
  .object({
    key: z.string(),
    latest_version: z.number(),
    latest_status: z.string(),
    created_at: z.string(),
  })
  .passthrough();
export type PromptSummary = z.infer<typeof promptSummarySchema>;

// 完整模板（GET /prompts/:key/versions 中的版本条目）。
// key 是主键，无 id 字段；status ∈ draft|published|archived。
export const promptTemplateSchema = z
  .object({
    key: z.string(),
    tenant_id: z.string().nullable().optional(),
    version: z.number(),
    content: z.string(),
    status: z.string(),
    content_hash: z.string(),
    created_by: z.string().optional().default(''),
    created_at: z.string(),
  })
  .passthrough();
export type PromptTemplate = z.infer<typeof promptTemplateSchema>;

// A/B 绑定（GET /prompts/bindings）。
// stable_version_id / canary_version_id 语义 = 版本对象的 content_hash（内容寻址）。
export const promptBindingSchema = z
  .object({
    key: z.string(),
    scope: z.string(), // tenant:<id> | agent:<id>
    stable_version_id: z.string(),
    canary_version_id: z.string().optional().default(''),
    traffic_percent: z.number().optional().default(0),
  })
  .passthrough();
export type PromptBinding = z.infer<typeof promptBindingSchema>;

export const PROMPT_STATUS_LABELS: Record<string, string> = {
  draft: '草稿',
  published: '已发布',
  archived: '已归档',
};

export const PROMPT_STATUS_COLORS: Record<string, string> = {
  draft: 'default',
  published: 'green',
  archived: 'orange',
};
